ssl/statem/extensions.c

```diff
@@ -834,6 +834,11 @@ static int final_server_name(SSL *s, unsigned int context, int sent,
         ret = s->session_ctx->ext.servername_cb(s, &altmp,
                                        s->session_ctx->ext.servername_arg);
 
+    if (!sent) {
+        OPENSSL_free(s->session->ext.hostname);
+        s->session->ext.hostname = NULL;
```

Comment:
This change is intended to fix a bug with early data, a TLS 1.3-only feature.
However, if I am reading things correctly, we can also run this code for pre-1.3 handshakes, and since we do not send SNI in the ServerHello in some cases (e.g., when resuming), we end up freeing a field in the SSL_SESSION object for a resumed session, i.e., one that is still in cache (for non-ticket cases).  This is, or should be, forbidden, because cached sessions can be in use by multiple threads simultaneously -- for the TLS 1.3 case we fix this by always creating a new SSL_SESSION object for each connection, but TLS 1.2 and older still have the old behavior.

One could perhaps argue that we should not be storing the "intermediate-state" (i.e., during handshake processing) hostname in the SSL_SESSION at all, and should instead be using the SSL's ext.hostname field.  We would then only populate the session once we know that we have negotiated an SNI value and are establishing a successful session, so the SSL_SESSION would remain read-only after (logical) creation.

Reply:
We are getting a double-free condition at this particular condition. We can get it to happen fairly readily, but don't know the exact cause (i.e. modifying a cached session).

Reply:
Hi,
https://tools.ietf.org/html/draft-ietf-tls-tls13-28#section-4.2.11 says :
"
In TLS 1.3, the SNI value is always explicitly specified in
   the resumption handshake, and there is no need for the server to
   associate an SNI value with the ticket.
"
So it seems that there is no need to zero ext.hostname.
Instead there is a need to check the existence of ext.hostname.

I also think that the draft should mention that SNI is required also at initial handshake.

Reply:
It is multi-threaded, and the current leading hypothesis is that this code's violation of the threading rules for ```s->session``` is causing the problems.
It is perhaps debatable whether we should be writing a hostname into the session during handshake processing but before we have decided to accept SNI, though, which would be a somewhat more invasive patch.

Reply:
> Is this a multi-threaded server?

Yes, this is the path we are currently going down; and agree with the SSL_SESSION write-access violation.

Reply:
Oops, somehow missed that @kaduk already said basically the same thing. :-)

> One could perhaps argue that we should not be storing the "intermediate-state" (i.e., during handshake processing) hostname in the SSL_SESSION at all, and should instead be using the SSL's ext.hostname field. We would then only populate the session once we know that we have negotiated an SNI value and are establishing a successful session, so the SSL_SESSION would remain read-only after (logical) creation.

This is what we do. Or rather, it's what we did when we stored the hostname in the `SSL_SESSION` at all. We've since stopped doing that. But note there is some weirdness here around being consistent whether, in a TLS 1.2 resumption, `SSL_get_servername` returns the value stored in the session or the value received on the connection, if you don't enforce that they match. That API is sometimes queried during and sometimes after the handshake.

Reply:
Indeed, there'd probably need to be some awful hack where ```SSL_get_servername``` can return the hostname from the SSL during the handshake and the one from the session otherwise.  Maybe that's more messiness than it's worth, and we should just go for the minimal change of only doing this free for non-cached session objects (i.e., non-resumptions and TLS 1.3 connections).

Reply:
Okay, I made #6378 to show what this approach looks like.